source/extensions/dynamic_modules/abi.h

```diff
@@ -22,13 +22,46 @@ extern "C" {
 // -----------------------------------------------------------------------------
 //
 // Types used in the ABI. The name of a type must be prefixed with "envoy_dynamic_module_type_".
+// Types with "_module_ptr" suffix are pointers owned by the module, i.e. memory space allocated by
+// the module. Types with "_envoy_ptr" suffix are pointers owned by Envoy, i.e. memory space
+// allocated by Envoy.
 
 /**
- * envoy_dynamic_module_type_abi_version represents a null-terminated string that contains the ABI
- * version of the dynamic module. This is used to ensure that the dynamic module is built against
- * the compatible version of the ABI.
+ * envoy_dynamic_module_type_abi_version_envoy_ptr represents a null-terminated string that
+ * contains the ABI version of the dynamic module. This is used to ensure that the dynamic module is
+ * built against the compatible version of the ABI.
+ *
+ * OWNERSHIP: Envoy owns the pointer.
+ */
+typedef const char* // NOLINT(modernize-use-using)
+    envoy_dynamic_module_type_abi_version_envoy_ptr;
+
+/**
+ * envoy_dynamic_module_type_http_filter_config_envoy_ptr is a raw pointer to
+ * the DynamicModuleHttpFilterConfig class in Envoy. This is passed to the module when
+ * creating a new in-module HTTP filter configuration and used to access the HTTP filter-scoped
+ * information such as metadata, metrics, etc.
+ *
+ * This has 1:1 correspondence with envoy_dynamic_module_type_http_filter_config_module_ptr in
+ * the module.
+ *
+ * OWNERSHIP: Envoy owns the pointer.
+ */
+typedef const void* // NOLINT(modernize-use-using)
+    envoy_dynamic_module_type_http_filter_config_envoy_ptr;
```

Comment:
I'm not sure I understand the purpose of passing void* here if the user has to have the actual C++ definition. Is this just to avoid any C++ leaking into the ABI itself?

Reply:
That's a valid question. The users (modules) do not need to understand the C++ definition, but instead users can just call the corresponding ABI functions with this opaque pointers. For example, one notable example about filter-config level ABI function is "creating metrics". Then possible function signature for that ABI would look like 

```c
envoy_dynamic_module_type_counter_metric_envoy_ptr envoy_dynamic_module_define_counter_metric(
    envoy_dynamic_module_type_http_filter_config_envoy_ptr envoy_filter_config_ptr,
    const char* name,
    int name_length,
);

```

where the module->Envoy ABI function has this opaque pointer as a first argument. Then on the Envoy side, we can provide the functionality

```c
envoy_dynamic_module_type_counter_metric_envoy_ptr envoy_dynamic_module_define_counter_metric(
    envoy_dynamic_module_type_http_filter_config_envoy_ptr envoy_filter_config_ptr,
    const char* name,
    int name_length,
) {
  const std::string_view name_view(static_cast<const char*>(name), name_length);
  auto scope = static_cast<DynamicModuleHttpFilterConfig*>(envoy_filter_config_ptr)->stat_scope_;
  Stats::StatNameManagedStorage storage(name_view, scope->symbolTable());
  Stats::StatName stat_name = storage.statName();
  Stats::Counter& counter = scope->counterFromStatName(stat_name);
  return &counter;
}
```

without leaking the implementation details of this envoy side "HttpFilterConfig" class into dynamic modules.

Apologies this is difficult to see the rationale behind this pointer passing without the concrete impl example like this.   Later when I add these ABI functions, then we will be able to see the concrete picture more clearly. Hope this helps!